text/0026-finality-signature-gossiping.org

```diff
@@ -0,0 +1,182 @@
+   |               | No                           | Yes         |
+   |               v                              v             |
+   |    +----------o------------+       +---------o--------+    |
+   |    |                       |       |                  |    |
+   |    |  Last Block Near      |       |    Get Block     |    |
+   +----+  Current Timestamp?   |       |    Using Hash    |    |
+   | No |                       |       |                  |    |
+   |    +----------+------------+       +---------o--------+    |
+   |               |                              |             |
+   |               | Yes                          |             |
+   |               v                              |             |
+   |       +-------o-------+                      +-------------+
+   |       |               |
+   |       |  Era Started  |
+   +-------+  Recently?    |
+     No    |               |
+           +-------+-------+
+                   |
+                   | Yes
+                   v
+           ,~~~~~~~o~~~~~~~.
+           :               :
+           :   Switch to   :
+           :   Highway     :
+           :   Consensus   :
+           :               :
+           `~~~~~~~~~~~~~~~'
+
+
+#+END_SRC
+
+#+RESULTS:
+[[file:images/0025/state-logic.svg]]
+
+** Reference-level explanation
+:PROPERTIES:
+:CUSTOM_ID: reference-level-explanation
+:END:
+
+The central place where program logic will change is in
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L365-L368][linear_chain_sync::LinearChainSync::fetch_next_block]].
+
+#+BEGIN_SRC rust
+  State::SyncingDescendants { .. } => {
+      let next_height = block_header.height() + 1;
+      fetch_block_at_height(effect_builder, peer, next_height)
+  }
+#+END_SRC
+
+Before the call to =fetch_block_by_hash=, the =LinearChainSync= process
+will need to have acquired a enough weight in signatures as discussed
+in [[Guide-level explanation][Guide-level explanation]].
+
+Here =fetch_block_at_height= will yield the block header /and/ the
+finality signatures. Moreover, updating =next_height= to
+=block_header.height() + 1= will happen only after enough weight of
+finality signatures has been observed.
+
+Byzantine nodes are a key concern for calculating the weight.
+
+Nodes must deal with a faulty validator when syncing.  A validator
+could make finality signatures with the same height (according to the
+block header) and different block hashes.  An approach to deal with
+this is as follows. First, introduce an enum:
+
+#+begin_src rust
+enum FaultyOrFetchedFinalitySignature {
+  /// A pair of finality signatures where:
+  /// - `bad_sig1.public_key == bad_sig2.public_key`
+  /// - `bad_sig1.block_height == bad_sig2.block_height`
+  /// - `bad_sig1.block_hash != bad_sig2.block_hash`
+  Faulty {
+    bad_sig1: FinalitySignature,
+    bad_sig2: FinalitySignature
+  },
+  Fetched(FinalitySignature)
```

Comment:
Maybe `Correct`/`Valid`/`Honest` would match better the `Faulty` state?

Reply:
It's not necessarily valid, because another finality signature could be gossiped and it would then be labelled as faulty.

Reply:
Until that happens it is valid. If another finality signature invalidates it then we will change enum variant to `Faulty`. Like a state machine.

Reply:
Less confusing for me to leave it as `Fetched`.  Does that not work for you?

Reply:
To clarify, it looks like we already have an enum `Observation` [in panorama][1] that is similar, so we should follow that convention.  There we used the term `Correct`.

[1]: https://github.com/goral09/casper-node-1/blob/0bbefb1c284784f0307661cdc7c33584ffded3ce/node/src/components/consensus/highway_core/state/panorama.rs#L24-L31

Reply:
Fixed here: https://github.com/CasperLabs/ceps/pull/26/commits/cc33b7063751e917c68c428feec34be1d3661ab0